src/darksend.cpp

```diff
@@ -1154,43 +1099,41 @@ bool CDarksendPool::SendDenominate(const std::vector<CTxIn>& vecTxIn, const std:
 bool CDarksendPool::UpdatePoolStateOnClient(PoolState nStateNew, int nEntriesCountNew, PoolStatusUpdate nStatusUpdate, PoolMessage nMessageID, int nSessionIDNew)
 {
     if(fMasterNode) return false;
-    if(nState == POOL_STATE_ERROR || nState == POOL_STATE_SUCCESS) return false;
 
-    SetState(nStateNew);
-    nEntriesCount = nEntriesCountNew;
+    // do not update state when mixing client state is one of these
+    if(nState == POOL_STATE_IDLE || nState == POOL_STATE_ERROR || nState == POOL_STATE_SUCCESS) return false;
 
     strAutoDenomResult = _("Masternode:") + " " + GetMessageByID(nMessageID);
 
-    if(nStatusUpdate != STATUS_SET_STATE) {
-        fLastEntryAccepted = nStatusUpdate;
-        if(nStatusUpdate == STATUS_REJECTED) {
-            SetState(POOL_STATE_ERROR);
-            strLastMessage = GetMessageByID(nMessageID);
-        }
+    // if rejected at any state
+    if(nStatusUpdate == STATUS_REJECTED) {
+        LogPrintf("CDarksendPool::UpdatePoolStateOnClient -- entry is rejected by Masternode\n");
+        UnlockCoins();
+        SetNull();
+        SetState(POOL_STATE_ERROR);
+        strLastMessage = GetMessageByID(nMessageID);
+        return true;
```

Comment:
`nLastTimeChanged` is changed at every _successful_ step, not only as a part of `CreateNewSession()` - in both `SetNull()` and `DoAuto()` if we talk about (re)starting a session. Also it times out only when `nState != POOL_STATE_IDLE`. Technically, time is set first `DoAuto()` _before_ switching from idle to queue, but it's two lines next to each other in the same function and I doubt it can cause issues. I think it won't hurt changing the order of instructions though for better readability/logic of handling them. Also `nLastTimeChanged` name itself doesn't say what it really does now, I'll fix that too.